src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -106,8 +107,45 @@ instance (Eq a, MultiplicativeMonoid a) => Package (ArithmeticCircuit a) where
     unpackWith f (ArithmeticCircuit c o) = ArithmeticCircuit c <$> f o
     packWith f = ArithmeticCircuit <$> foldMap acCircuit <*> f . fmap acOutput
 
-instance (Eq a, MultiplicativeMonoid a) => Symbolic (ArithmeticCircuit a) where
+type Arithmetic a = (WitnessField a, Eq a)
+
+instance Arithmetic a => Symbolic (ArithmeticCircuit a) where
     type BaseField (ArithmeticCircuit a) = a
+    symbolicF (ArithmeticCircuit c o) _ f = let (p, d) = runState (f o) c in withOutputs d p
+
+-------------------------------- MonadCircuit instance ------------------------------
+
+instance Arithmetic a => MonadCircuit Natural a (State (Circuit a)) where
+    newRanged upperBound witness = do
+        let s   = sources @a witness
+            -- | A wild (and obviously incorrect) approximation of
+            -- x (x - 1) ... (x - upperBound)
+            -- It's ok because we only use it for variable generation anyway.
+            p i = var i * (var i - fromConstant upperBound)
```

Comment:
We should think of a safer way to do it. `p` could be used to define a constraint `x \in {0, upperBound}`.

Reply:
Oh, definitely, didn't think about set inclusion constraints when writing this